Scripts for SUSE SLE HPC

petersatsuse opened the pull request:
This PR will add the SUSE SLE HPC 15 SP4  Product as a new tree to the project:
AZCOPY=10.16.1
HPCX=v2.12
MVAPICH2=2.3.6
OMPI=4.1.1
IMPI_2021=2021.7.0
CUDA=11.8.0
NVIDIA=520.61.05
AOCL=3.1.0
AOCL=3.1.0
INTEL_ONE_MKL=2022.2.0
NVIDIA-DOCKER=20.10.17_ce
DCGM=3.0.4
WAAGENT=2.2.49.2

abhamidipati0614 commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install_utils.sh:
```diff
@@ -0,0 +1,135 @@
+## SLES HPC ship with many HPC packages already, so no need to build it - simple install is enough
+#
+# as the default cloud image does only install parts of the HPC toolchain we use the pattern
+zypper install -y --type pattern hpc_libraries
```
I believe this should be a comment

petersatsuse replied:
no, what we have available in the product and what the SLE HPC images has pre-installed within the marketpalce image is different. The marketplace image is has only a minimal set of packages installed (just enough os principle).
So using a pattern (= a group/list of packages) to install our default setup is needed

abhamidipati0614 commented on tests/run-tests.sh:
```diff
@@ -283,6 +324,32 @@ then
+
+    CHECK_AOCL=1
+    CHECK_NV_PMEM=0
+    CHECK_NCCL=0
```
I believe nccl is being installed, any reason why it is off?

petersatsuse replied:
oops, i added nccl in an second changeset and forgot to enable the test.

abhamidipati0614 commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/README.md:
```diff
@@ -0,0 +1,57 @@
+
+The SLE HPC 15 SP4 includes optimizations and recommended configurations to deliver optimal performance,
+consistency, and reliability.
+As an enterprise distribution, we provide stable long term support and many certifications with vendors.
```
Can you please also add that SLE HPC 15 SP4 is maintained by SUSE?

abhamidipati0614 commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install_nvidiagpudriver.sh:
```diff
@@ -0,0 +1,37 @@
+
+NVIDIA_VERSION=$(rpm -q --qf="%{VERSION}" cuda-drivers)
+$COMMON_DIR/write_component_version.sh "NVIDIA" ${NVIDIA_VERSION}
+
```
Is it installed with dkms support? nvidia related packages are dependent on kernel and will break with kernel updates. Does SUSE behave similarly? If so, I think it is a good idea to install with dkms and hold the nvidia and related packages (that depend on kernel) from updating

abhamidipati0614 commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install_utils.sh:
```diff
@@ -0,0 +1,135 @@
+    kernel-macros \
+    binutils \
+    kernel-headers \
+    kernel-source \
```
Is it a good idea to hold kernel from updating as most of the hpc packages are dependent on kernel?

petersatsuse replied:
We do not change the kernel and the kernel api within a ServicePack. This is one of the differences and the promise of a Enterprise Linux distribution vs. community distributions - a stable kABI.

abhamidipati0614 commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install_docker.sh:
```diff
@@ -0,0 +1,24 @@
+systemctl --now enable docker
+
+# Test to run the hw container
+# docker run --rm hello-world
```
This is added in run-tests.sh

petersatsuse replied:
was only a comment how to test the docker install - not needed at all here.

abhamidipati0614 reviewed (commented):
### Additional thoughts
- Can you please share which build of SLE 15 is this tested on? If it is available on azure, can you please point the urn? I tried _**SUSE:sles-15-sp4:gen2:2022.11.04**_ but it fails at installing `lua-lmod`
- Can you please share if you have any performance numbers and metrics?
- Can you please also add a status badge to the main page? - it will help reflect the image health

petersatsuse commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install_utils.sh:
```diff
@@ -0,0 +1,135 @@
+    libtool \
+    rdma-core-devel
+
+# the ibdev2netdev is only in the external mellanox package, so we do not have it with inbox drivers
```
ok, will move it.

petersatsuse commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install_utils.sh:
```diff
@@ -0,0 +1,135 @@
+AZCOPY_URL="https://aka.ms/downloadazcopy-v10-linux"
+
+# not accessible anymore at this blob - its private now
+# wget https://azcopyvnextrelease.blob.core.windows.net/release20210920/azcopy_linux_se_amd64_10.12.2.tar.gz
```
ok, will adapt it to the new versions and handling

petersatsuse commented on tests/run-tests.sh:
```diff
@@ -251,6 +290,7 @@ then
     CHECK_NCCL=1
 elif [[ $distro == "Ubuntu 18.04" ]]
 then
+#TODO - check if this path get reached as the function above does only return the version
```
no problem

petersatsuse commented on suse/sle-hpc-15.x/sle-hpc-15-sp4/install.sh:
```diff
@@ -0,0 +1,67 @@
+$COMMON_DIR/network-tuning.sh
+
+# copy test file
+$COMMON_DIR/copy_test_file.sh
```
thx

petersatsuse commented:
> ### Additional thoughts
> 
>     * Can you please share which build of SLE 15 is this tested on? If it is available on azure, can you please point the urn? I tried _**SUSE:sles-15-sp4:gen2:2022.11.04**_ but it fails at installing `lua-lmod`

Its clear, as you used the "normal" SUSE Linux Enterprise Server (SLES) and not the HPC product  ( sle-hpc-15-sp4 ), where we ship additional HPC applications and environments, e.g lua ;)

The URN should be 
   "suse:sle-hpc-15-sp4..." 
the gen and image version should not matter as our image is identical for gen1 and gen2 and the image release-version will change every 3 month when we refresh the image.

Our images have a lifecyle as we want to provide a) a always secure image to start with b) less update needed 

>     * Can you please share if you have any performance numbers and metrics?

As i do not have access to all needed azure instances, i only had some output from the test script, e.g

#---------------------------------------------------
# Benchmarking PingPong 
# #processes = 2 
#---------------------------------------------------
       #bytes #repetitions      t[usec]   Mbytes/sec
            0         1000         0.36         0.00
            1         1000         0.35         2.82
            2         1000         0.35         5.64
            4         1000         0.36        11.16
            8         1000         0.35        22.71
           16         1000         0.35        45.26
           32         1000         0.42        75.40
           64         1000         0.46       140.47
          128         1000         0.48       264.30
          256         1000         0.55       461.99
          512         1000         0.55       932.28
         1024         1000         0.78      1319.77
         2048         1000         1.09      1883.89
         4096         1000         1.34      3053.21
         8192         1000         2.08      3941.82
        16384         1000         2.55      6417.85
        32768         1000         2.94     11147.04
        65536          640         4.97     13177.46
       131072          320         9.29     14115.00
       262144          160        13.58     19305.54
       524288           80        34.10     15373.50
      1048576           40        69.53     15080.35
      2097152           20       141.03     14870.69
      4194304           10       243.85     17200.22

> Can you please also add a status badge to the main page? - it will help reflect the image health

It do not have access to the interal testing for the badge...it's still in discussion how to handle that.

abhamidipati0614 commented:
which azure sku is the benchmarking performed?

petersatsuse commented:
> @petersatsuse which Azure SKU is the benchmarking performed on?

As mentioned in the question right before the numbers all is based on "suse:sle-hpc-15-sp4..."
Our gen1 and gen2 images are the same, only azure has here the limitation from the instance.
The "date" should not take any role as we maintain as Enterprise distribution stability.

Do not remember the instance type, as i still do not have access to the one used at MSFT.

abhamidipati0614 commented:
@petersatsuse  I tried the scripts on an NDv4 A100 machine with SUSE:sle-hpc-15-sp4:gen2:latest and I see the following error during NCCL installation

```
+ make -j src.build
make -C src build BUILDDIR=/tmp/nccl-2.15.1-1/build
make[1]: Entering directory '/tmp/nccl-2.15.1-1/src'
NVCC_GENCODE is -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=comput=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90,code90
Grabbing   include/nccl_net.h                  > /tmp/nccl-2.15.1-1/build/include/nccl_net.h
Generating nccl.h.in                           > /tmp/nccl-2.15.1-1/build/include/nccl.h
Generating nccl.pc.in                          > /tmp/nccl-2.15.1-1/build/lib/pkgconfig/nccl.pc
Compiling  init.cc                             > /tmp/nccl-2.15.1-1/build/obj/init.o
Compiling  channel.cc                          > /tmp/nccl-2.15.1-1/build/obj/channel.o
Compiling  bootstrap.cc                        > /tmp/nccl-2.15.1-1/build/obj/bootstrap.o
Compiling  enqueue.cc                          > /tmp/nccl-2.15.1-1/build/obj/enqueue.o
Compiling  transport.cc                        > /tmp/nccl-2.15.1-1/build/obj/transport.o
Compiling  proxy.cc                            > /tmp/nccl-2.15.1-1/build/obj/proxy.o
Compiling  group.cc                            > /tmp/nccl-2.15.1-1/build/obj/group.o
Compiling  debug.cc                            > /tmp/nccl-2.15.1-1/build/obj/debug.o
Compiling  net.cc                              > /tmp/nccl-2.15.1-1/build/obj/net.o
Compiling  misc/cudawrap.cc                    > /tmp/nccl-2.15.1-1/build/obj/misc/cudawrap.o
Compiling  misc/nvmlwrap.cc                    > /tmp/nccl-2.15.1-1/build/obj/misc/nvmlwrap.o
Compiling  misc/ibvwrap.cc                     > /tmp/nccl-2.15.1-1/build/obj/misc/ibvwrap.o
Compiling  misc/gdrwrap.cc                     > /tmp/nccl-2.15.1-1/build/obj/misc/gdrwrap.o
Compiling  misc/utils.cc                       > /tmp/nccl-2.15.1-1/build/obj/misc/utils.o
Compiling  misc/argcheck.cc                    > /tmp/nccl-2.15.1-1/build/obj/misc/argcheck.o
Compiling  misc/socket.cc                      > /tmp/nccl-2.15.1-1/build/obj/misc/socket.o
Compiling  misc/shmutils.cc                    > /tmp/nccl-2.15.1-1/build/obj/misc/shmutils.o
Compiling  misc/profiler.cc                    > /tmp/nccl-2.15.1-1/build/obj/misc/profiler.o
Compiling  misc/param.cc                       > /tmp/nccl-2.15.1-1/build/obj/misc/param.o
Compiling  misc/strongstream.cc                > /tmp/nccl-2.15.1-1/build/obj/misc/strongstream.o
Compiling  transport/p2p.cc                    > /tmp/nccl-2.15.1-1/build/obj/transport/p2p.o
Compiling  transport/shm.cc                    > /tmp/nccl-2.15.1-1/build/obj/transport/shm.o
Compiling  transport/net.cc                    > /tmp/nccl-2.15.1-1/build/obj/transport/net.o
Compiling  transport/net_ib.cc                 > /tmp/nccl-2.15.1-1/build/obj/transport/net_ib.o
Compiling  transport/net_socket.cc             > /tmp/nccl-2.15.1-1/build/obj/transport/net_socket.o
Compiling  transport/coll_net.cc               > /tmp/nccl-2.15.1-1/build/obj/transport/coll_net.o
Compiling  collectives/sendrecv.cc             > /tmp/nccl-2.15.1-1/build/obj/collectives/sendrecv.o
Compiling  collectives/all_reduce.cc           > /tmp/nccl-2.15.1-1/build/obj/collectives/all_reduce.o
Compiling  collectives/all_gather.cc           > /tmp/nccl-2.15.1-1/build/obj/collectives/all_gather.o
Compiling  collectives/broadcast.cc            > /tmp/nccl-2.15.1-1/build/obj/collectives/broadcast.o
Compiling  collectives/reduce.cc               > /tmp/nccl-2.15.1-1/build/obj/collectives/reduce.o
Compiling  collectives/reduce_scatter.cc       > /tmp/nccl-2.15.1-1/build/obj/collectives/reduce_scatter.o
Compiling  graph/topo.cc                       > /tmp/nccl-2.15.1-1/build/obj/graph/topo.o
Compiling  graph/paths.cc                      > /tmp/nccl-2.15.1-1/build/obj/graph/paths.o
Compiling  graph/search.cc                     > /tmp/nccl-2.15.1-1/build/obj/graph/search.o
Compiling  graph/connect.cc                    > /tmp/nccl-2.15.1-1/build/obj/graph/connect.o
Compiling  graph/rings.cc                      > /tmp/nccl-2.15.1-1/build/obj/graph/rings.o
Compiling  graph/trees.cc                      > /tmp/nccl-2.15.1-1/build/obj/graph/trees.o
Compiling  graph/tuning.cc                     > /tmp/nccl-2.15.1-1/build/obj/graph/tuning.o
Compiling  graph/xml.cc                        > /tmp/nccl-2.15.1-1/build/obj/graph/xml.o
make[2]: Entering directory '/tmp/nccl-2.15.1-1/src/collectives/device'
Compiling  enhcompat.cc                        > /tmp/nccl-2.15.1-1/build/obj/enhcompat.o
NVCC_GENCODE is -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=comput=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90,code90
Generating rules                               > /tmp/nccl-2.15.1-1/build/obj/collectives/device/Makefile.rules
NVCC_GENCODE is -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=comput=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90,code90
nvcc fatal   : Unsupported gpu architecture 'compute_35'
nvcc fatal   : Unsupported gpu architecture 'compute_35'
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/sendrecv.dep] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/all_reduce.dep] Error 1
nvcc fatal   : Unsupported gpu architecture 'compute_35'
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/all_gather.dep] Error 1
nvcc fatal   : Unsupported gpu architecture 'compute_35'
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/broadcast.dep] Error 1
nvcc fatal   : Unsupported gpu architecture 'compute_35'
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/reduce.dep] Error 1
nvcc fatal   : Unsupported gpu architecture 'compute_35'
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/reduce_scatter.dep] Error 1
nvcc fatal   : Unsupported gpu architecture 'compute_35'
nvcc fatal   : Unsupported gpu architecture 'compute_35'
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/functions.dep] Error 1
make[2]: *** [Makefile:53: /tmp/nccl-2.15.1-1/build/obj/collectives/device/onerank_reduce.dep] Error 1
make[2]: Leaving directory '/tmp/nccl-2.15.1-1/src/collectives/device'
make[1]: *** [Makefile:57: /tmp/nccl-2.15.1-1/build/obj/collectives/device/colldevice.a] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/tmp/nccl-2.15.1-1/src'
make: *** [Makefile:25: src.build] Error 2
```

Seems like it is expecting an explicit GPU arch specification.
`*Cloud you pls. have a look at the open questions and suggestions too?`
If you mean by sunrpc_tcp_settings comment, I will check with @jithinjosepkl about it. Otherwise, please clarify.

petersatsuse commented:
> @petersatsuse I tried the scripts on an NDv4 A100 machine with SUSE:sle-hpc-15-sp4:gen2:latest and I see the following error during NCCL installation
> 
> ```
> + make -j src.build
> make -C src build BUILDDIR=/tmp/nccl-2.15.1-1/build
> make[1]: Entering directory '/tmp/nccl-2.15.1-1/src'
> NVCC_GENCODE is -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=comput=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90,code90
> Grabbing   include/nccl_net.h                  > /tmp/nccl-2.15.1-1/build/include/nccl_net.h
...
> Generating rules                               > /tmp/nccl-2.15.1-1/build/obj/collectives/device/Makefile.rules
> NVCC_GENCODE is -gencode=arch=compute_35,code=sm_35 -gencode=arch=compute_50,code=sm_50 -gencode=arch=compute_60,code=sm_60 -gencode=arch=comput=sm_61 -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90,code90
> nvcc fatal   : Unsupported gpu architecture 'compute_35'
> nvcc fatal   : Unsupported gpu architecture 'compute_35'
..
> 
> Seems like it is expecting an explicit GPU arch specification.

As far as i found out, the error shows that the Kepler virtual architecture (compute_35) is not supported by the used Nvida Card and CUDA and NCCL combination.

By default, NCCL is compiled for all (supported) architectures. To accelerate the compilation and reduce the binary size, consider redefining NVCC_GENCODE (defined in makefiles/common.mk) to only include the architecture of the target platform :
  $ make -j src.build NVCC_GENCODE="-gencode=arch=compute_90,code=sm_80"

The makefile mentioned above  does include compute_35 for CUDA < 12. We use CUDA 11.8 here.
Not sure why you dont see this within your alma tree as it uses similar (bit older) versions.

  ifeq ($(shell test "0$(CUDA_MAJOR)" -lt 12; echo $$?),0)  
    # SM35 is deprecated from CUDA12.0 onwards
    CUDA8_GENCODE += -gencode=arch=compute_35,code=sm_35
  endif

I do not have access to such a machine at the moment, so i cant try it myself (did a SR to get one)

 `*Cloud you pls. have a look at the open questions and suggestions too?` If you mean by sunrpc_tcp_settings comment, I will check with @jithinjosepkl about it. Otherwise, please clarify.
Yes, 
but i in addition i was not sure if you had read all my answers, as i had more than one question. But it seems i closed them mistakenly.

petersatsuse commented:
After talking to our NVIDIA expert its more clear. With CUDA 12 they dropped the Kepler architecture:
As 12 shows up in the SLES Nvidia repository, it get automatic selected as its the newest version.

As i dont have at the moment any instance with NV available, could you try to replace in "install_nvidiagpudriver.sh" at line 27, "cuda" with "cuda-11-8" and give it a new try.

petersatsuse commented:
@abhamidipati0614
Could you pls. provide me the instance type you used? 
I only have access to a Standard_HB120-16rs_v2 and the build of the nccl test goes through without error - see below

...
Compiling  all_reduce.cu                       > /var/lib/waagent/run-command/download/0/azhpc-images-suse/suse/sle-hpc-15.x/sle-hpc-15-sp4/nccl-tests/build/all_reduce.o
Compiling  common.cu                           > /var/lib/waagent/run-command/download/0/azhpc-images-suse/suse/sle-hpc-15.x/sle-hpc-15-sp4/nccl-tests/build/common.o
Linking  /var/lib/waagent/run-command/download/0/azhpc-images-suse/suse/sle-hpc-15.x/sle-hpc-15-sp4/nccl-tests/build/all_reduce.o > /var/lib/waagent/run-command/download/0/azhpc-images-suse/suse/sle-hpc-15.x/sle-hpc-15-sp4/nccl-tests/build/all_reduce_perf
Compiling  all_gather.cu   
...
make[1]: Leaving directory '/var/lib/waagent/run-command/download/0/azhpc-images-suse/suse/sle-hpc-15.x/sle-hpc-15-sp4/nccl-tests/src'
+ popd
/var/lib/waagent/run-command/download/0/azhpc-images-suse/suse/sle-hpc-15.x/sle-hpc-15-sp4
+ mv nccl-tests /opt

petersatsuse commented:
@abhamidipati0614 sorry for the late answer - had to take care of different things.
 I found during additional testing on more instances some versions issues, like the one with cuda, as some repositories got many new versions since the intital PR, which do not match in some cases.
 
I still do not have access/quota at bigger instances, so my testings are on ND40rs_v2 (V100) where everything is fine, including a successfull run-tests,
Secont is a NC24ads_A100_v4 (A100) where run-test does not work due to missing IB, but nccl test got build without errors. It has the A100 like your NDv4 A100.

If you like, i could provide the output from the run-tests directly via email

petersatsuse commented:
@abhamidipati0614  - I finally got a NDv4 A100 (ND96asr_v4) and was no able to see your error.

In my run all go through without error. The only thing additonally needed, was to enable the fabricmanager service, otherwise the NCCL tests with 8 nodes from run-rests showed an error.
Please have a look - i can provide you the logs by email

petersatsuse commented:
wonder a bit about the error, as this codepath had no change since month...and in my run from last week it did not show up. If i run it today it does...Strange. Will dig into it ... and sure resolve the merge conflicts which show up in the meantime

abhamidipati0614 commented:
thank you for fixing the issues. Here's another one from run-tests.sh (please also commit run-tests.sh with execute permissions)

/opt/azurehpc/test/run-tests.sh --mofed-lts false

Detected distro: SUSE Linux Enterprise High Performance Computing 15 SP4
ofed inbox driver does miss the ofed_info tool
/opt/azurehpc/test/run-tests.sh: line 504: lspci: command not found
*** Error - IB device not found!

abhamidipati0614 commented:
@petersatsuse - we usually don't call the run-tests.sh with sudo. This is due to the fact that MPI runs wouldn't like to be run as root. We call run-tests.sh as an admin user other than root and wouldn't need sudo permissions. This was never a problem for other distros.

_running as root would lead to the following*
```
mpirun has detected an attempt to run as root.

Running as root is *strongly* discouraged as any mistake (e.g., in
defining TMPDIR) or bug can result in catastrophic damage to the OS
file system, leaving your system in an unusable state.

We strongly suggest that you run mpirun as a non-root user.

You can override this protection by adding the --allow-run-as-root option
to the cmd line or by setting two environment variables in the following way:
the variable OMPI_ALLOW_RUN_AS_ROOT=1 to indicate the desire to override this
protection, and OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1 to confirm the choice and
add one more layer of certainty that you want to do so.
We reiterate our advice against doing so - please proceed at your own risk.
```

petersatsuse commented:
@abhamidipati0614 - yes, nothing should run as root if its not for managing the machine - its always a good advice :)

At SUSE we have lspci in the path /sbin, so in system binaries. The path is not in the PATH for normal users, even if they can execute the command, as there are scenarios where you can crash a machine with it. Second it exposes quite a lot of internals of the system ... 

I will add a new variable which got filled with the full path of the lspci command for all the distri's, as its different in all distros. So than with that, we will have it universal.

I still wonder a bit why there is " --allow-run-as-root " in some mpirun commands as parameter in run-tests.sh. Should they  not get removed, as it would allow exactly that?

petersatsuse commented:
@abhamidipati0614 - within the SUSE variable set section from run-tests.sh i added the two needed directories to the PATH. I also got aware that the user running the script need to be in the group "video". Not sure how you solved that in the other distri's - i added a note to the REAME

abhamidipati0614 commented:
@petersatsuse  Thank you for the commits. It works now. Only change I had to make was to start the NVIDIA fabric manager on A100. I see the sku customization service is not setup. Do you also want to leverage this service which can tune the guest based on  the hardware. Refer - https://github.com/Azure/azhpc-images/blob/master/common/setup_sku_customizations.sh

petersatsuse commented:
@abhamidipati0614

I looked at the sku customizations and its nice, but i have some issues with the setup_nvme.sh script.

The handling of md arrays is not universal, as 
a) the mdadm has several options where the config file could live, which could be different in distris and customer setups 
b) you simply overwrite or delete all md devices with UUIDs - which maybe belong to other md arrays, 
c) update-initramfs is imho Debian specific, RH and SUSE use dracut
d) not sure why you need update-initramfs at all, as this is imho only neccesary for md boot devices
e) /mnt is with newer azure agents/cloud-init ephemeral, but as the script recreate it every time it does not hurt

I would like to first move to a simpler maintainable configuration way, which i prepare right now for a next PR on top of the SUSE tree. It would also a way for the other distri setups to be more easy to move to other minor or major versions of all included parts

abhamidipati0614 commented:
@petersatsuse that sounds good. Please sync the branch and we can merge this.

abhamidipati0614 commented:
/azp run
